### PR TITLE
Improve scrolling experience in browse mode

### LINKED ISSUE #6382: Visual issues in browse mode
There seem to be some visual issues in browse mode in several browsers. @feerrenrut, you might be interested in this in particular. I use nnvaccess.org for my examples.

## Focus changes
As expected, when arrowing down a page in browse mode, the focus is continuously updated to reflect the last selected focusable element. So, when you navigate through heading level 2 news with the arrow keys, the focus is set to the link MORE TESTIMONIALS, since that's the link which received focus last.

However, undesirable behaviour rises when I set the virtual cursor to the top of the page and navigate to the news heading with quick navigation commands (either h or 2). The focus is still set to the NV Access link at the top of the page, so there is no visual indication that we scrolled past a bunch of links we skipped due to quick nav.

I'd like to propose that, when using forward quick navigation commands to navigate to an unfocusable element, the focus is moved to the last focusable element above the navigator object (below in the case of backwards navigation). In that case, people with vision at least will have an idea where the NVDA user is on the screen. Furthermore, the visual behaviour of browse mode moving will be identical in both quicknav and arrowing cases.

## Inconsistency between browsers
Scrolling behaviour with browse mode differs between browsers and thus is inconsistent. I think we should strive to consistent behaviour whenever possible. I prefer the Internet Explorer approach for bringing the navigator object into view.

### Firefox/Chrome
When navigating with tab (browser command), the focus is set properly and the focused element is visible at the very bottom of the page. When I use tab to navigate to the first news link, jump back to the news heading with shift+h and than down arrow back to the news link, the visual result is equal. When I navigate the page by heading, the heading is always one of the last visible elements on the screen, making it quite unusable for sighted users who follow the nnon-visual person's actions. Normal behaviour would be that one reads the text below a heading after selecting the desired heading, but the text below the heading is invisible for the most part.

## Internet Explorer
When navigating with tab, the focus is set properly and the focused element is visible at the bottom of the page. However, when I use tab to navigate to the first news link, jump back to the news heading with shift+h and than down arrow back to the news link, the visual result differs entirely. The particular link is visible at the top of the page. When I navigate the page by heading, the heading is always one of the first visible elements on the screen, which is quite nice and desirable behaviour from a visual person's perspective. However, it is inconsistent with the browser's behaviour.

### LINKED ISSUE #7075: Content doesn't scroll in Edge
In Microsoft Edge, when NVDA is reading simple text (with no focusable objects like links, buttons, etc. inside it), the web page does not scroll to show the currently read text. So, if the user reads a long segment of simple text from top to bottom, Edge will show only the part which currently fits vertically on the screen, no matter that NVDA has passed reading that part and it is reading way down the text. If NVDA encounters a focusable element on the page (e.g. a button, a link, etc.), Edge will scroll the view to show that focusable element/object on the screen. Is this a limitation in Edge or a bug in NVDA?

Steps to reproduce it:
1. Start NVDA.
2. Open Microsoft Edge and load a page in it, which contains a long segment (it may be of separate paragraphs) of simple text (with no focusable elements/objects like links, buttons, etc. in it). A good example is the following: [https://www.nvaccess.org/about/our-story](url)
3. Start reading the "Our Story" article with the arrow keys.

Expected results:
NVDA should scroll the web page, so it will constantly show on the screen what it is currently reading.

Actual results:
The web page does not scroll.

Test environment:
- Operating system: Windows 10 Pro (build 15063.138), 64-bit, in Bulgarian with all locale settings set to "Bulgarian".
- NVDA version: next-13994,744a2d39.
- Microsoft Edge version: 40.15063.0.0.
- Microsoft EdgeHTML version: 15.15063.
- Processor: Intel Core i5-2320 at 3.00GHz.
- RAM Memory: 8.00GB.
- Graphics: Palit GeForce GT 630, 1024MB dedicated memory + 3071MB shared system memory, desktop resolution set to 1920x1080.
- Sound Card: Realtek ALC662 at Intel Cougar Point PCH - High Definition Audio Controller.

### PR DESCRIPTION
**Pull request by LeonarddeR**
### Link to issue number:
Fixes #7075 
Closes #6382 
Should fix #904 

### Summary of the issue:
There are several scrolling issues in browse mode, particularly in Edge, where the screen doesn't scroll along with browse mode.

### Description of how this pull request fixes the issue:
This pr does the following:

1. It brings a scrollIntoView method to textInfos. There are several implementations for this, particularly UIA, IA2, Acrobat Object Model and Word object model.
2. Browse mode calls scrollIntoView on the text info, and if that fails, it tries the current object as a fallback. However, on the object level, we do not support the allignToTop parameter whereas on the text info level, we do.
3. For most implementations, it makes sure that text will only scroll when the browse mode cursor gets off screen.

### Testing performed:
Many tests have been performed by @k-kolev1985 as part of #7075, but I have changed some stuff since then. In any case, this requires testing by sighted people.

### Known issues with pull request:
In internet Explorer, we do not support the onlyWhenVisible parameter for scrollIntoView. This is because if we do, navigating with browse mode in Internet Explorer slows down significantly.

### Change log entry:
* Bug fixes
    + NVDA will now properly scroll text in browse mode in Microsoft Edge, Firefox, Chrome, Adobe Reader and Microsoft Word. (#6382, ##7075)
    + When using say all in browse mode, text on screen now scrolls along with the text being read. (#904)

### CODE REVIEW

**Comment by leonardder** [comment]
@codeofdusk: It might be good to see what this scrolling code does in UIA consoles. Note that you will still have to trigger it manually though, as the review cursor commands do not yet scroll.

**Comment by leonardder** [comment]
[Here](https://ci.appveyor.com/api/buildjobs/0f1p1kdktol8tnix/artifacts/output%2Fnvda_snapshot_try-UIAScroll-18047%2C07f49c80.exe) is a try build.

**Comment by codeofdusk** [comment]
> @codeofdusk: It might be good to see what this scrolling code does in UIA consoles. Note that you will still have to trigger it manually though, as the review cursor commands do not yet scroll.

Trying to scroll a `textInfo` into view that is beyond the first visible range (by calling its `scrollIntoView` method in the Python console) makes review unusable until the caret moves/new text is added. Knowing if text actually scrolls visually would require sighted input. Cc @feerrenrut

**Comment by leonardder** [comment]
@codeofdusk: Could you elaborate on what you mean with `makes review unusable`?

**Comment by leonardder** [comment]
@JulienCochuyt: You might be interested in this pr, since it will drastically improve scrolling behavior in web browsers for sighted users.

**Comment by Qchristensen** [comment]
Using the About NV Access page (the address has changed from the original issue, but I think it's still the same page, and it does meet the criteria of having a large block of non-linked text): https://www.nvaccess.org/about-nv-access/

With NVDA 2019.2, both Firefox (68.0.2) and Chrome (76.0.3809.100), both scroll as I press the down arrow to move through text.  Neither they, nor Edge scroll when using say all however.

Using this PR build, Edge now scrolls when I down arrow through the document, where it doesn't with 2019.2.  Neither of the other browsers scroll with down arrow now, and none of them scroll during say all.

@feerrenrut can you confirm either my findings, or that I've gone mad?

**Comment by leonardder** [comment]
So in summary, in the current situation there is a regression for 
firefox and Chrome? That's certainly not what should happen.

**Comment by leonardder** [comment]
Ah, I found it. Will try to come up with an appropriate fix. It also seems that using NVDA+delete is appropriate to test this nonvisually.

**Comment by leonardder** [comment]
It turns out that IAccessibleTextObject.scrollSubStringTo fails in some cases, that is, it returns S_OK but still does not scroll. Pretty annoying, since it indeed causes a regression. This means that we could only scroll at the object level.

**Comment by MarcoZehe** [comment]
Is this a bug in Firefox, or do you see this in Chrome as well?

**Comment by leonardder** [comment]
@Qchristensen Would you be able to retest this with the newest build?

note that say all is expected not to work. This is covered in #9937. I could consider putting this in a separate pull request, but I had to change some logic as part of #9937, so it made sense to do that now.

**Comment by leonardder** [comment]
@MarcoZehe wrote:
> Is this a bug in Firefox, or do you see this in Chrome as well?

I've seen the issue in both, but I might have a clarification for it. I"m probably calling scrollSubStringTo on a text range for which the object is classified off screen. Now, when the object is completely out of view, I scroll the object instead.

I realise though that this can still cause some screen bouncing when the object is very large. I should probably scroll the object first and do another scroll for the text.

**Comment by leonardder** [comment]
@JulienCochuyt As a fully blind user, it is pretty difficult for me to bring this in a state that actually works how I want it to be. Would you mind taking a look? I'd imagine @accessolutions could benefit from this work as much as @babbagecom will do.

File: source/NVDAObjects/IAccessible/MSHTML.py [hunk 1]
```diff
@@ -911,13 +920,22 @@ def _get_rowCount(self):
...
-			self.HTMLNode.scrollInToView()
+			node.scrollIntoView(alignToTop)
 		except (COMError,NameError):
-			pass
+			log.debugWarning("HTML node scrollIntoView failed", exc_info=True)
+			raise NotImplementedError
+
+	def scrollIntoView(self):
+		if self.HTMLNode:
+			self._scrollHTMLNodeIntoView(self.HTMLNode, True)
```
Comment (JulienCochuyt) [inline source/NVDAObjects/IAccessible/MSHTML.py]: Shouldn't this line return to avoid raising `NotImplementedError` if the test succeeded?

```suggestion
			return self._scrollHTMLNodeIntoView(self.HTMLNode, True)
```

File: source/NVDAObjects/IAccessible/__init__.py [hunk 2]
```diff
@@ -357,6 +357,27 @@ def _iterTextWithEmbeddedObjects(self, withFields, formatConfig=None):
...
+	def _scrollIntoView(self, alignToTop: bool = True):
+		scrollType = (
+			IAccessibleHandler.IA2_SCROLL_TYPE_TOP_EDGE
+			if alignToTop
+			else IAccessibleHandler.IA2_SCROLL_TYPE_BOTTOM_EDGE
+		)
+		try:
+			if self.obj.hasIrrelevantLocation:
+				# The complete object is out of view.
+				# In these cases, scrollSUbStringTo is known to fail.
```
Comment (JulienCochuyt) [inline source/NVDAObjects/IAccessible/__init__.py]: Small typo
```suggestion
				# In these cases, scrollSubStringTo is known to fail.
```

**Comment by JulienCochuyt** [comment]
> @JulienCochuyt As a fully blind user, it is pretty difficult for me to bring this in a state that actually works how I want it to be. Would you mind taking a look? I'd imagine @accessolutions could benefit from this work as much as @BabbageCom will do.

@leonardder: Sure, I'll try to dig into this, but please note I probably won't have time to thoroughly review this big scope until at least third week of October.

Could you please summarize your intended target behavior, just to be sure we are on the same track here?
Especially regarding "alignToTop": In which cases exactly did you consider it necessary, and does "alignToTop=False" mean "alignToBottom" or just "ensureVisible"? Is there such a consideration as "alignToCenter" ?

As a side note, re Adobe browse mode, did you consider #5881 and #6673?

**Comment by leonardder** [comment]
> @leonardder: Sure, I'll try to dig into this, but please note I probably won't have time to thoroughly review this big scope until at least third week of October.

That's fine with me.

> Could you please summarize your intended target behavior, just to be sure we are on the same track here?

I had tree primary targets:

1. When the browse mode virtual caret scrolls into a piece on the page that is invisible, the screen should scroll in such a way that the virtual caret always remains visible on screen.
2. Scrolling behavior should be uniform accross browsers, at least Firefox, Chrome and Edge.
3. Scrolling behavior should be somewhat similar to scrilling behavior in normal documents, i.e. in Word, Notepad, etc. IN any case, it should be intuitive and what people would expect to happen when scrolling.

> Especially regarding "alignToTop": 

There is some discrepancy between accessibility API implementations regarding scrolling behavior.

* IA2: Allows you to scroll to top, bottom, left, right and anywhere for both objects and sub strings
* UIA: ALlows you to scroll anywhere for objects, top and bottom for sub strings
* MSHTML: allows you to scroll to top or bottom for objects
* Word: allows you to scroll to top or bottom

> In which cases exactly did you consider it necessary,

If I remember correctly, scrollIntoView does always scroll for MSHTML, Word and UIA text ranges. So, we need allignToTop to decide whether it has to scroll to the top or bottom of the view port.

> and does "alignToTop=False" mean "alignToBottom" or just "ensureVisible"? 

The former indeed.

> Is there such a consideration as "alignToCenter" ?

It is not supported 

> As a side note, re Adobe browse mode, did you consider #5881 and #6673?

Not yet.

**Comment by leonardder** [comment]
I think this should also fix #5881. Not sure about #6673, though, it might even make it worse.

**Comment by JulienCochuyt** [comment]
> about #6673, though, it might even make it worse.

This would goes beyond my imagination, as the current effect is already quite drastic, especially when you simultaneously attempt to circumvent it using the mouse scroll wheel.

**Comment by JulienCochuyt** [comment]
@leonardder thanks for all the details.
Thus, I guess we are attempting here to alignToTop whenever moving from a visible object to an upper invisible object and alignToBottom when moving to a lower invisible object.
I consider here the origin of the move rather than the attempted direction on purpose: Eg. depending on the document structure, hitting the down arrow might move the virtual caret somewhere up.

**Comment by leonardder** [comment]
@JulienCochuyt: sorry for me taking some time to get back to this.

> Thus, I guess we are attempting here to alignToTop whenever moving from a visible object to an upper invisible object and alignToBottom when moving to a lower invisible object.

Exactly that.

> I consider here the origin of the move rather than the attempted direction on purpose: Eg. depending on the document structure, hitting the down arrow might move the virtual caret somewhere up.

Could you elaborate with an example? Hitting down arrow while the caret is at the bottom should idially scroll the new text into view, leaving the caret at the bottom of the view port.

**Comment by JulienCochuyt** [comment]
> Could you elaborate with an example? Hitting down arrow while the caret is at the bottom should idially scroll the new text into view, leaving the caret at the bottom of the view port.

If you are browsing content presented in columns by any means, when moving down from the last line of a given column, your destination is the first line of the next column, thus alignToTop in this case.
A specific example is when the canva defines a region between the main content and the footer, visually anchored at the upper right corner of the main content.

**Comment by leonardder** [comment]
Ah, yes, of course. That's a very interesting one I  hadn't considered.

**Comment by leonardder** [comment]
I wrote this pull request on behalf of @babbagecom. As I'm leaving @babbagecom after the 29th of November, I can no longer afford maintaining this pr other than applying very basic review actions. If this pull request requires major changes, they will have to be applied by someone else, e.g. @JulienCochuyt, @sjfbol or whoever else is willing to take it.

**Comment by JulienCochuyt** [comment]
@leonardder As promised, I'll take care of this one (unless of course if @sjfbol is willing to)
I must concentrate on other projects right now with the upcoming of 2019.3, but this is definitely in our pipeline.
I guess you can leave this PR open for reference for now. When I'll return on polishing your code, I'll mark the new PR as superseding this one and closing it.

**Comment by leonardder** [comment]
It's unlikely that @sjfbol will jump on this pr in the near future, so it would probably best if you do as suggested. I will Finnish this off with fixing the merge conflicts and then I'll leave this open.

**Comment by lukaszgo1** [comment]
As in the current Alphas fast browse mode became the default what effect would it have on this PR? Is it even possible to have working scrolling for partially sighted users and do not move focus in browse mode at the same time?

**Comment by JulienCochuyt** [comment]
@lukaszgo1 wrote:
> As in the current Alphas fast browse mode became the default what effect would it have on this PR? Is it even possible to have working scrolling for partially sighted users and do not move focus in browse mode at the same time?

As promised, I've started work on taking back this PR, but it requires a huge lot of testing for corner cases.
Theoretically, fast browse mode should not interfere as the focused object can always be safely scrolled out.
Still, thanks for drawing attention on this use case to include in the test suite.

**Comment by codeofdusk** [comment]
@JulienCochuyt Any updates on the testing of this PR?

**Comment by leonardder** [comment]
As it seems that this pr hasn't gained much attention by potential takers. Furthermore, when one takes over a pr, it is a common pattern that the old pr is closed in favour of a new one.

I'm closing this pull request to clean up the backlog, happy for anyone to reopen a pr based on this work if desired.

**Comment by Adriani90** [comment]
I think it would be very beneficial to have this work done, especially for visually impaired people and when working together with sighted people.
Given Internet explorer is discontinued, we don't need to set priority on IE anymore.
However, MsHTML is still being used in several places but nevertheless this PR would be a significant improvement to the current situation anyway.
